---
title: "Assignment 6 RNAseqPart 1"
output: html_document
---

__Name:__ Koki Sasagawa  
__Student ID:__ 999054646

**Always include the code or commands used to answer the exercises**

**When done, knit an html and push both this .Rmd and the .html to your github.  Start an issue requesting grading**.  No PDF needed.
```{r}
#Converting from GFF to GTF file format
library(rtracklayer)
gff <- import.gff("~/Assignments/Assignment_6_Sasagawa.Koki/Brapa_reference/Brapa_reference/Brapa_gene_v1.5.gff")
gff #take a look

#create a column "gene_id" that contains the gene name for every entry
gff$gene_id <- ifelse(is.na(gff$ID),gff$Parent,gff$ID)

export(gff,"Brapa_reference/Brapa_gene_v1.5.gtf",format="gtf")
```
In order to ask if reads are differentially expressed between IMB211 vs R500 or treatments we need to know how many reads were sequenced from each gene.
```{r}
library(Rsubread)
readCounts <- featureCounts(
  files=c("../Assignment_5_Sasagawa.Koki/tophat_out-IMB211_All_A01_INTERNODE.fq/accepted_hits_A01.bam",
    "../Assignment_5_Sasagawa.Koki/tophat_out-R500_All_A01_INTERNODE.fq/accepted_hits_A01.bam"),
  annot.ext="./Brapa_reference/Brapa_gene_v1.5.gtf", 
  isGTFAnnotationFile=TRUE,
  GTF.featureType="CDS", # This depends on GTF file.  Often it would be "exon"
  GTF.attrType="gene_id"
  )
```
**Exercise 1**  
Read the help file for feature counts.  Be sure to look at the section "Value" where it describes the output.  
__a__. Provide a line of code that displays the counts of the first 6 genes.  (It is not very interesting because the first genes in the file are on chromosome A03 (strange numbering...) and our bam file only has counts from A01...  )  

> The function featureCounts assigns mapped sequencing reads to genomic features.

```{r}
#Display the counts of the first 6 genes
head(readCounts$counts, n = 6)
```
__b__. The gene `Bra011030` is on chromosome A01.  What are its read counts in the two files?  
```{r}
readCounts$counts["Bra011030",]
```
> In IMB211 & R500, on chromosome A01 read counts for gene Bra011030 are 1462 & 1930.

__c__. What percentage of reads (from each file) were assigned to a gene?  What percentage were unassigned because they were not located in a gene (aka "Feature")?  
```{r}
#Total number of reads
#IMB211
a <- sum(readCounts$stat[,2])
#R500
a1 <- sum(readCounts$stat[,3])

#The total number of assigned reads
#IMB211
b <- sum(readCounts$stat[1,2])
#R500
b1 <- sum(readCounts$stat[1,3])

#Check
sprintf("Total number of IMB211 reads: %i", a)
sprintf("Total number of R500 reads: %i", a1)
sprintf("Total number of IMB211 assigned reads: %i", b)
sprintf("Total number of R500 assigned reads: %i", b1)

#Total number of unassigned reads
#IMB211
y <- a - b
#R500
y1 <- a1 - b1

#Check
sprintf("Total number of IMB211 unassigned reads: %i", y)
sprintf("Total number of R500 unassigned reads: %i", y1)

#Total % of unassigned reads
#IMB211
z <- (y / a) * 100
#R500
z1 <- (y1 / a1) * 100

#Total % of assigned reads
#IMB211
w <- (b / a) * 100
#R500
w1 <- (b1 / a1) * 100

#Answer
sprintf("The total percent of unassigned reads for IMB211 is %f and for assigned reads is %f", z, w)
sprintf("The total percent of unassigned reads for R500 is %f and for assigned reads is %f", z1, w1)
```

__d__. What are 2 possible reasons why there are reads that cannot be assigned to a gene?  

> The unmatched read can be due to PCR error in amplification of cDNA. Foriegn reads can also arise from contamination from outside sources.

**Exercise 2**  
Move the downloaded data to your current working directory.

__a__. Create a new object in R called `counts.data` with either the leaf data or the internode data.  
```{r}
counts.data <- read.delim("~/Assignments/Assignment_6_Sasagawa.Koki/gh_internode_counts.tsv", header = TRUE)
```
__b__. Check to make sure that the data looks as expected.  (What do you expect and how do you confirm?  Show your commands and output.)

> I expect to see the read counts for each gene in each of the samples. 

```{r}
#Run head command to get a sample view of the dataframe to see if the information I want is present.
head(counts.data)
```
```{r}
#Remove unmatched reads
counts.data <- counts.data[rownames(counts.data)!="*",]
counts.data[is.na(counts.data)] <- 0
```

**Exercise 3**  
The column names are too long.  Use the `sub()` command to remove the ".1_matched.merged.fq.bam" suffix from each column name.  Although it doesn't matter in this case, using the argument "fixed=TRUE" is a good idea because "." is a wildcard character.

> sub(pattern, replacement, x, fixed)
The pattern is the user specified pattern to search for. Setting fixed=TRUE looks for a pattern without using regular expressions. Replacement is a character string to replace the patterns. X is the character vector to search for the pattern. In this case, we are specifying colnames(counts.data) as our vector as we want to look for the pattern specifically in the column header. 

```{r}
colnames(counts.data) <- sub(pattern = ".1_matched.merged.fq.bam", replacement = "", x = colnames(counts.data), fixed = TRUE)
head(counts.data)
```
**Exercise 4**  
__a__. Make a histogram of counts for each of the samples.  
```{r}
library(ggplot2)
library(reshape2)
library(ggthemes)

counts.data.melt <- melt(counts.data[,colnames(counts.data)], id.vars = NULL)

graph1 <- ggplot(data = counts.data.melt, aes(x = value, fill = variable)) 
graph1 + geom_histogram() + facet_wrap(~variable) + theme_dark() + xlab("# of Counts Per Gene") + ylab("# Of Genes") + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```
__b__. Is the data normally distributed? Apply an appropriate transformation if needed and make a new set of histograms. 
__Hint 1__: _see the use of the `melt()` function in the Rice-SNP lab_.  __Hint 2__: _You can transform the axes in ggplot by adding `scale_x_log10()` or `scale_x_sqrt()` to the plot.  One of these should be sufficient if you need to transorm, but for other ideas see the [Cookbook for R page](http://www.cookbook-r.com/Graphs/Axes_%28ggplot2%29/#axis-transformations-log-sqrt-etc)_.  

> The data is not normally distributed. It is getting skewed to the left due to the high occurence of genes with low read counts in each of the samples

```{r}
#install.packages('ggthemes', dependencies = TRUE)
library(ggplot2)
library(ggthemes)
graph1 <- ggplot(data = counts.data.melt, aes(x = value, fill = variable)) 
graph1 + geom_histogram() + facet_wrap(~variable) + scale_y_log10() + scale_x_log10() + theme_dark() + xlab("# of Counts Per Gene") + ylab("# Of Genes") + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```
```{r}
#retain genes with > 10 reads in >= 3 samples
#For each row, rowSums combines the number of reads in each column together.
counts.data <- counts.data[rowSums(counts.data > 10) >= 3,]
```
**Exercise 5**  
We expect that read counts, especially from biological replicates, will be highly correlated.  Check to see if this is the case using the `pairs()` function and the `cor()` function.  Explain what each of these functions does and comment on the results.  __Important Hint:__ _`pairs` is slow on the full dataset.  Try it on the first 1,000 genes.  Do you need to transform to make the pairs output more meaningful?_

> pairs() function produces a scatterplot, creating a correlation matrix plot in between parameters in the dataset. It is useful for creating a visual representation of correlation. It takes arguments x (the coordinates of points given as numeric columns of matrix or dataframe), formula, data, subset, na.action (function that indicates instructions for when data contains NA), labels (name of variables) to list a few.

> cor() function produces numerical correlational values (ranging from 0 - 1). It takes x (Matrix or dataframe), use (Specify handling of missing data), and method (Specify type of correlation, options are pearson, spearman, or kendall).

```{r}
#Create a new dataframe for the first 1000 genes. 
first1k <- counts.data[1:1000,]

#Produce a scatterplot using pairs(). Compare read counts for first 1000 genes between the replicates
pairs(first1k, lower.panel = panel.smooth, upper.panel = NULL)

#Run cor() to obtain correlational values
cor(counts.data)
```

```{r}
#The previous scatterplot does not have clean correlation 
#Perform a log10() transformation for better correlation
pairs(log10(first1k), lower.panel = panel.smooth, upper.panel = NULL)
```

> Comparing the two methods, cor() runs much faster, and produces numbers that correspond to the correlation (0-absolutely no correlation / 1-perfect correlation) between each treatment and their read counts for each gene. pairs() takes significantly longer to run even with the smaller dataframe, but produces a scatterplot matrix which is a nice way to visually represent correlation. Each point on the graph is a gene which is plotted by its read counts along the x-axis and y-axis that correspond to 2 different samples. For correlational studies, cor() may be more effective when working with larger datasets. 

__Exercise 6__  
Discuss the MDS plot.  Does it give you confidence in the experiment or cause concern?

```{r}
sample.description <- data.frame(
  sample=colnames(counts.data),
  
  #This next line searches for IMB211 or R500 in the colnames of counts.data and returns anything that matches
  #In this way we can extract the genotype info.
  gt=regmatches(colnames(counts.data),regexpr("R500|IMB211",colnames(counts.data))),
  
  #Now we use the same method to get the treatment
  trt=regmatches(colnames(counts.data),regexpr("NDP|DP",colnames(counts.data)))
  )

# Now we can paste the trt and gt columns together to give a group identifier
sample.description$group <- paste(sample.description$gt,sample.description$trt,sep="_")

# set the reference treatment to "NDP"
sample.description$trt <- relevel(sample.description$trt,ref="NDP")

sample.description
```

```{r}
#Use edgeR function calcNormFactors() to calculate normalization factors
library(edgeR)
dge.data <- DGEList(counts=counts.data, group=sample.description$group)
dim(dge.data) 
dge.data <- calcNormFactors(dge.data, method = "TMM")
dge.data$samples # look at the normalization factors
```

Statistically significant differential expression is assessed based on variance within and between treatments. We can partition the variance between replicate RNA samples into two sources: technical variance and biological variance.Technical variance usually decreases as the number of total gene count increases, but biological variance will not change even if total library size increases. 

coefficient of variation (CV) is calculated as standard deviation divided by the mean. 
```{r}
#Use biological coefficient of variation to calculate distance between samples based off their gene expression. 
plotMDS(dge.data, method = "bcv") 
```

Discuss the MDS plot. Does it give you confidence in the experiment or cause concern?

> The MDS plot visually represents the similarity of each sample based off BCV. We expect biological replicants to be highly correlated/similar. Looking at the plot we see that biological replicants are indeed grouping up with each other. We can also observe that the 4 different treatments are distinctly separate from eachother and grouping up in the 4 corners of the plot. As these are different cultivars under different treatments the result gives us confidence in our experiment. 

__Exercise 7__  
To get a graphical idea for what the normalization does, make box plots of the count data for each sample before and after normalization. Discuss the effect of normalization.

__Hint 1__: _log2 transform the counts before plotting.  Add a value of "1" before log2 transforming to avoid having to take the log2 of 0.
__Hint 2__: If you donâ€™t want to bother with melting before going to ggplot, you can just use the boxplot() function and feed it the (transformed) matrix directly.

```{r}
library(reshape2)
library(ggplot2)
library(ggthemes)

#Extract normalized data
counts.data.normal <- cpm(dge.data)

#log2 transform counts
counts.data.log <- log2(counts.data + 1)
counts.data.normal.log <- log2(counts.data.normal + 1)

#melt the data to proper format
counts.data.normal.log.melt <- melt(counts.data.normal.log[,colnames(counts.data.normal)], id.vars = NULL)
counts.data.log.melt <- melt(counts.data.log[,colnames(counts.data.log)], id.vars = NULL)

#Generate the boxplot
ggplot(counts.data.log.melt, aes(x = variable, y = value, fill = variable)) + geom_boxplot() + theme_dark() + xlab("Sample") + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))

ggplot(counts.data.normal.log.melt, aes(x = counts.data.normal.log.melt$Var2, y = value, fill = counts.data.normal.log.melt$Var2)) + geom_boxplot() + xlab("Sample") + theme_dark() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```

> Normalization allows us to identify differentially expressed genes by allowing us to compare the different samples together. The overall average is calculated to create a reference point that each of the samples can be compared relatively to. This allows us to compare different samples expression levels relative to the average, allowing us to detect differentially expressed genes.

__Exercise 8__  
Dispersion describes how "squeezed" or "broad" a distribution is. In a negative binomial distribution the BCV is the square root of the dispersion of the negative binomial distribution. Calculate the gene-specific (tagwise) dispersion to allow genes whose expression is consistent between replicates to be ranked more highly than genes that are not.

```{r}
design <- model.matrix(~gt+trt,data = sample.description)
rownames(design) <- sample.description$sample
design
```
```{r}
#First the overall dispersion
dge.data <- estimateGLMCommonDisp(dge.data,design,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data <- estimateGLMTrendedDisp(dge.data,design)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data <- estimateGLMTagwiseDisp(dge.data,design)

#We can examine this with a plot
plotBCV(dge.data)
```
```{r}
#fit a generalized linear model (GLM) for each gene
#This is the full model that contains coefficients for genotype and treatment.
fit <- glmFit(dge.data, design)
```
Look for differentially expressed genes by comparing the full model to a reduced model with fewer coefficients. 
```{r}
#find genes that are differentially expressed in R500 vs IMB211
gt.lrt <- glmLRT(fit,coef = "gtR500")
```
View the top 10 differentially expressed genes
```{r}
topTags(gt.lrt) # the top 10 most differentially expressed genes
```
```{r}
#summarize the number of differentially expressed genes 
summary(decideTestsDGE(gt.lrt,p.value=0.01)) #This uses the FDR.
```
The -1 row is the number of down regulated genes in R500 relative to IMB211. 0 is the same, and +1 is the upregulated genes.
```{r}
#Extract genes with a FDR < 0.01 (could also use 0.05)
#Inf means infinity and is important to extract all the genes. topTags is the only way to extract data. 
DEgene.gt <- topTags(gt.lrt,n = Inf)$table[topTags(gt.lrt,n = Inf)$table$FDR<0.01,]

#save to a file
write.csv(DEgene.gt,"DEgenes.gt.csv")
```

```{r}
#Define plot bar chart function
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  require(reshape2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- merge(tmp.data,sample.description,by.x="row.names",by.y="sample")
  tmp.data <- melt(tmp.data,value.name="log2_cpm",variable.name="gene")
  pl <- ggplot(tmp.data,aes(x=gt,y=log2_cpm,fill=trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

```{r}
# A single gene
plotDE("Bra009785",dge.data,sample.description)

#top 9 genes
plotDE(rownames(DEgene.gt)[1:9],dge.data,sample.description)
```

__a__.  Find all genes differentially expressed in response to the DP treatment (at a FDR < 0.01).  
```{r}
#find genes that are differentially expressed in response to DP treatment
#compare all the 1 to 0
trt.lrt <- glmLRT(fit,coef = "trtDP")

#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes.trt <- topTags(trt.lrt,n = Inf)$table[topTags(trt.lrt,n = Inf)$table$FDR<0.01,]

write.csv(DEgenes.trt,"DEgenes.trt.csv")
```
__b__.  How many genes are differentially expressed?
```{r}
#summarize the number of differentially expressed genes 
summary(decideTestsDGE(trt.lrt,p.value=0.01)) #This uses the FDR.
```
> The total number of differentially expressed genes is 619

__c__.  Make a plot of the top 9
```{r}
#top 9 genes
plotDE(rownames(DEgenes.trt)[1:9],dge.data,sample.description)
```

__Exercise 9__:  
Examining the plots of the top gene regulated by treatment you will notice that several of the genes have a different response to treatment in IMB211 as compared to R500. We might be specifically interested in finding those genes. How could we find them? We look for genes whose expression values are fit significantly better by a model that includes a genotype X treatment term.
```{r}
design.interaction <- model.matrix(~gt*trt,data = sample.description)
rownames(design.interaction) <- sample.description$sample
design.interaction

#First the overall dispersion
dge.data.2 <- estimateGLMCommonDisp(dge.data,design.interaction,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data.2 <- estimateGLMTrendedDisp(dge.data,design.interaction)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data.2 <- estimateGLMTagwiseDisp(dge.data,design.interaction)

#fit a generalized linear model (GLM) for each gene
#This is the full model that contains coefficients for genotype and treatment.
fit2 <- glmFit(dge.data.2, design.interaction)
```

__a__. Repeat the dispersion estimates and model fit but with the new model.  Show code.
```{r}
#find genes that are differentially expressed in response to DP treatment in R500 and IMB211
gt.trt.lrt <- glmLRT(fit2,coef = "gtR500:trtDP")

#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgenes.gt.trt <- topTags(gt.trt.lrt,n = Inf)$table[topTags(gt.trt.lrt,n = Inf)$table$FDR<0.01,]
```
__b__. How many genes show a significantly different response to treatment in IMB211 as compared to R500?  Save these genes to a file.  
```{r}
#summarize the number of differentially expressed genes 
summary(decideTestsDGE(gt.trt.lrt,p.value=0.01)) #This uses the FDR.

write.csv(DEgenes.gt.trt,"DEgenes.gt.trt.csv")
```

> The total number of differentially expressed genes is 254.

__c__. Make a plot of the top 9 genes that have a significantly different response to treatment in IMB211 as compared to R500.  
```{r}
#top 9 genes
plotDE(rownames(DEgenes.gt.trt)[1:9],dge.data.2,sample.description)
```